File: oneflow/user/ops/adaptive_pool.cpp
/*
Copyright 2020 The OneFlow Authors. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
#include "oneflow/core/framework/framework.h"
#include "oneflow/user/ops/nn_util.h"

namespace oneflow {

namespace {

Maybe<void> InferTensorDesc(user_op::InferContext* ctx) {
  const user_op::TensorDesc* first_in_desc = ctx->TensorDesc4ArgNameAndIndex("in", 0);
  std::vector<int64_t> output_size = ctx->Attr<std::vector<int64_t>>("output_size");
  DimVector out_dim_vec = first_in_desc->shape().dim_vec();
  int h = out_dim_vec[2];
  int w = out_dim_vec[3];
  if(output_size.size() >= 1){
    h = output_size[0]; //h
    w = output_size[0];
    if(output_size.size() == 2){
      h = output_size[1]; //w
    }
  }
  out_dim_vec[2] = h;
  out_dim_vec[3] = w;
  user_op::TensorDesc* out_desc = ctx->TensorDesc4ArgNameAndIndex("out", 0);
  *out_desc->mut_shape() = Shape(out_dim_vec);
  *out_desc->mut_is_dynamic() = *ctx->IsDynamic4ArgNameAndIndex("x", 0);
}

Maybe<void> FwGetSbpFn(user_op::SbpContext* ctx) {
  const user_op::TensorDesc& tensor = ctx->LogicalTensorDesc4InputArgNameAndIndex("x", 0);
  // only for nchw
  FOR_RANGE(int64_t, i, 0, std::min(2, (int)tensor.shape().NumAxes())) {
    ctx->NewBuilder().Split(user_op::OpArg("x", 0), i).Split(user_op::OpArg("y", 0), i).Build();
  }
  return Maybe<void>::Ok();
}

REGISTER_USER_OP("adaptive_avg_pool2d")
    .Input("in")
    .Attr<std::vector<int64_t>>("output_size")
    .Output("out")
    .SetTensorDescInferFn(InferTensorDesc)
    .SetGetSbpFn(FwGetSbpFn);

REGISTER_USER_OP("adaptive_avg_pool2d_grad")
    .Input("x")
    .Input("dy")
    .Attr<std::vector<int64_t>>("output_size")
    .Output("dx")
    .SetTensorDescInferFn(InferTensorDesc)
    .SetGetSbpFn(FwGetSbpFn);

REGISTER_USER_OP_GRAD("adaptive_avg_pool2d").SetBackwardOpConfGenFn([](user_op::BackwardOpConfContext* ctx) {
  const auto adaptive_avg_pool2d_grad_op_name = ctx->FwOp().op_name() + "_grad";
  ctx->DefineOp(adaptive_avg_pool2d_grad_op_name, [&ctx](user_op::BackwardOpBuilder& builder) {
    return builder.OpTypeName("adaptive_avg_pool2d_grad")
        .InputBind("x", ctx->FwOp().input("in", 0))
        .InputBind("dy", ctx->FwOp().output_grad("out", 0))
        .Attr("output_size", ctx->FwOp().attr<double>("scale"))
        .Output("dx")
        .Build();
  });
  ctx->FwOp().InputGradBind(user_op::OpArg("in", 0),
                            [&ctx, &adaptive_avg_pool2d_grad_op_name]() -> const std::string& {
                              return ctx->GetOp(adaptive_avg_pool2d_grad_op_name).output("dx", 0);
                            });
});

}  // namespace

}  // namespace oneflow

File: python/oneflow/compatible/single_client/nn/modules/adaptive_pool.py
"""
Copyright 2020 The OneFlow Authors. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""
from oneflow.compatible import single_client as flow
from oneflow.compatible.single_client.nn.module import Module


class AdaptiveAvgPool2d(Module):
    """Applies a 2D adaptive average pooling over an input signal composed of several input planes.

    The output is of size H x W, for any input size.
    The number of output features is equal to the number of input planes.

    Args:
        output_size: the target output size of the image of the form H x W.
                     Can be a tuple (H, W) or a single H for a square image H x H.
                     H and W can be either a ``int``, or ``None`` which means the size will
                     be the same as that of the input.

    For example:

    .. code-block:: python

        >>> import numpy as np
        >>> import oneflow.compatible.single_client.experimental as flow
        >>> import oneflow.compatible.single_client.experimental.nn as nn
        >>> flow.enable_eager_execution()

        >>> m = nn.AdaptiveAvgPool2d((5,7))
        >>> input = flow.Tensor(np.random.randn(1, 64, 8, 9))
        >>> output = m(input)
        >>> output.size()
        flow.Size([1, 64, 5, 7])

        >>> m = nn.AdaptiveAvgPool2d(7)
        >>> input = flow.Tensor(np.random.randn(1, 64, 10, 9))
        >>> output = m(input)
        >>> output.size()
        flow.Size([1, 64, 7, 7])

        >>> m = nn.AdaptiveAvgPool2d((None, 7))
        >>> input = flow.Tensor(np.random.randn(1, 64, 10, 9))
        >>> output = m(input)
        >>> output.size()
        flow.Size([1, 64, 10, 7])

    """

    def __init__(self, output_size) -> None:
        super().__init__()
        self.output_size = output_size
        self._op = (
            flow.builtin_op("adaptive_avg_pool2d")
            .Input("x")
            .Attr("output_size", [])
            .Output("y")
            .Build()
        )

    def forward(self, x):
        new_output_size = []
        assert len(x.shape) == 4
        if isinstance(self.output_size, int):
            new_output_size.append(self.output_size)
            new_output_size.append(self.output_size)
        elif isinstance(self.output_size, tuple):
            new_output_size = list(self.output_size)
            if self.output_size[0] is None:
                new_output_size[0] = x.shape[2]
            if self.output_size[1] is None:
                new_output_size[1] = x.shape[3]
        else:
            raise NotImplementedError("output_size param wrong, please check!")
        new_output_size = tuple(new_output_size)
        assert (
            new_output_size[0] <= x.shape[2]
        ), f"output_size param wrong, please check!"
        assert (
            new_output_size[1] <= x.shape[3]
        ), f"output_size param wrong, please check!"
        return self._op(x, output_size=new_output_size)[0]


if __name__ == "__main__":
    import doctest

    doctest.testmod(raise_on_error=True)
